kernel/kernel-impl/src/main/java/org/sakaiproject/user/impl/JpaPreferencesService.java

```diff
@@ -0,0 +1,200 @@
+import org.sakaiproject.user.api.Preferences;
+import org.sakaiproject.user.api.PreferencesEdit;
+import org.sakaiproject.user.api.UserDirectoryService;
+import org.sakaiproject.user.api.repository.PreferenceRepository;
+import org.sakaiproject.user.api.model.Preference;
+import org.sakaiproject.util.StorageUtils;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class JpaPreferencesService extends BasePreferencesService {
+
+    @Setter
+    private PreferenceRepository preferenceRepository;
+
+    @Override
+    protected Storage newStorage() {
+        log.info("Creating new JPA-based preferences storage implementation");
+        return new JpaStorage();
+    }
+    
+    // Abstract method implementations for dependencies
+    private final MemoryService m_memoryService = null;
+    @Override
+    protected MemoryService memoryService() { return m_memoryService; }
+
+    private final ServerConfigurationService m_serverConfigurationService = null;
+    @Override
+    protected ServerConfigurationService serverConfigurationService() { return m_serverConfigurationService; }
+
+    private final EntityManager m_entityManager = null;
+    @Override
+    protected EntityManager entityManager() { return m_entityManager; }
+
+    private final SecurityService m_securityService = null;
+    @Override
+    protected SecurityService securityService() { return m_securityService; }
+
+    private final FunctionManager m_functionManager = null;
+    @Override
+    protected FunctionManager functionManager() { return m_functionManager; }
+
+    private final SessionManager m_sessionManager = null;
+    @Override
+    protected SessionManager sessionManager() { return m_sessionManager; }
+
+    private final EventTrackingService m_eventTrackingService = null;
+    @Override
+    protected EventTrackingService eventTrackingService() { return m_eventTrackingService; }
+
+    private final UserDirectoryService m_userDirectoryService = null;
+    @Override
+    protected UserDirectoryService userDirectoryService() { return m_userDirectoryService; }
+    
+    // Added for backward compatibility with SqlService
+    private final SqlService m_sqlService = null;
+    protected SqlService sqlService() { return m_sqlService; }
+    
+    /**
+     * Final initialization, once all dependencies are set.
+     */
+    @Override
+    public void init() {
+        try {
+            super.init();
+            
+            log.info("init(): JPA-based preference storage ready");
+        } catch (Exception t) {
+            log.warn("init(): ", t);
+        }
+    }
+    
+    /**
+     * Returns to uninitialized state.
+     */
+    @Override
+    public void destroy() {
+        super.destroy();
+        log.info("destroy(): JPA-based preference storage shutdown");
+    }
+
+    @lombok.RequiredArgsConstructor
+    protected class JpaStorage implements Storage {
```

Comment:
this is strange, looks like a cross between JPA and Sakai's traditional storage?